go/cmd/dolt/commands/sql.go

```diff
@@ -164,52 +164,52 @@ func (cmd SqlCmd) Exec(ctx context.Context, commandStr string, args []string, dE
 		}
 	}
 
-	origRoot := root
-
-	saveMessage := apr.GetValueOrDefault(messageFlag, "")
-	saveName := apr.GetValueOrDefault(saveFlag, "")
-
 	_, forceBatchMode := apr.GetValue(batchFlag)
 
-	query, queryOK := apr.GetValue(queryFlag)
-	if !queryOK {
-		savedQueryName, exOk := apr.GetValue(executeFlag)
-
-		if exOk {
-			sq, err := dsqle.RetrieveFromQueryCatalog(ctx, root, savedQueryName)
+	if query, queryOK := apr.GetValue(queryFlag); queryOK {
+		saveMessage := apr.GetValueOrDefault(messageFlag, "")
+		saveName := apr.GetValueOrDefault(saveFlag, "")
 
-			if err != nil {
-				return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
-			}
+		return execQuery(ctx, dEnv, root, query, saveName, saveMessage, forceBatchMode, format, usage)
+	}
 
-			query = sq.Query
-			queryOK = true
+	if savedQueryName, exOk := apr.GetValue(executeFlag); exOk {
+		sq, err := dsqle.RetrieveFromQueryCatalog(ctx, root, savedQueryName)
 
-			cli.PrintErrf("Executing saved query '%s':\n%s\n", savedQueryName, query)
+		if err != nil {
+			return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
 		}
-	}
 
-	if !queryOK {
-		if apr.Contains(listSavedFlag) {
-			hasQC, err := root.HasTable(ctx, doltdb.DoltQueryCatalogTableName)
+		cli.PrintErrf("Executing saved query '%s':\n%s\n", savedQueryName, sq.Query)
+		return execQuery(ctx, dEnv, root, sq.Query, "", "", forceBatchMode, format, usage)
+	}
 
-			if err != nil {
-				verr := errhand.BuildDError("error: Failed to read from repository.").AddCause(err).Build()
-				return HandleVErrAndExitCode(verr, usage)
-			}
+	if apr.Contains(listSavedFlag) {
+		hasQC, err := root.HasTable(ctx, doltdb.DoltQueryCatalogTableName)
 
-			if !hasQC {
-				return 0
-			}
+		if err != nil {
+			verr := errhand.BuildDError("error: Failed to read from repository.").AddCause(err).Build()
+			return HandleVErrAndExitCode(verr, usage)
+		}
 
-			query = "SELECT * FROM " + doltdb.DoltQueryCatalogTableName
-			queryOK = true
+		if !hasQC {
+			return 0
 		}
+
+		query := "SELECT * FROM " + doltdb.DoltQueryCatalogTableName
+		return execQuery(ctx, dEnv, root, query, "", "", forceBatchMode, format, usage)
 	}
 
+	return execQuery(ctx, dEnv, root, "", "", "", forceBatchMode, format, usage)
+}
+
+func execQuery(ctx context.Context, dEnv *env.DoltEnv, root *doltdb.RootValue, query, saveName, saveMessage string, forceBatchMode bool, format resultFormat, usage cli.UsagePrinter) int {
```

Comment:
This should be three separate functions invoked from the top level Exec function, one for each major code path.

execQuery should run a single query and exit
execBatchModeQueries should run batch mode
execShell should run the shell